### PR TITLE
drawer: new component

### PR DESCRIPTION
**Pull request by KaidenR**
### What You're Solving

- Adds `drawer` component

### Design Decisions

- Added support for light theme but since the `row` component does not yet support it, the colors are still of when they are used together.

### How to Verify

- View component storybook

### CODE REVIEW

File: packages/drawer/src/react/index.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+)
+
+const RowContainerWrapper = glamorous.div(
+  ({ themeName }) => ({
+    borderTop: `1px solid ${themeName === themeNames.dark ? gray04 : gray01 }`
+  })
+)
+const RowContainer = (props, context) => (
+  <RowContainerWrapper themeName={context.themeName}>
+    <Div marginTop={-1} padding={`0 57px 0 ${core.layout.spacingMedium}`}>
```
Comment (jaketrent) [inline packages/drawer/src/react/index.js]: What's the magic 57px for?

File: packages/drawer/src/react/index.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    : {}
+  )
+)
+
+const RowContainerWrapper = glamorous.div(
+  ({ themeName }) => ({
+    borderTop: `1px solid ${themeName === themeNames.dark ? gray04 : gray01 }`
+  })
+)
+const RowContainer = (props, context) => (
```
Comment (jaketrent) [inline packages/drawer/src/react/index.js]: Why the design choice to tie Row to the drawer like this?  Could it be solved in a more general way?

File: packages/drawer/src/react/index.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+  get isControlledByProps() {
+    return this.props.isOpen !== undefined && this.props.isOpen !== null
+  }
+  get isOpen() {
+    return this.isControlledByProps ? this.props.isOpen : this.state.isOpen
+  }
+
+  isClickOnDrawerBase(evt) {
+    return !evt.target.closest('a, button')
```
Comment (jaketrent) [inline packages/drawer/src/react/index.js]: `closest` has no IE support.  Is this being polyfilled?  Is there another solution?  Would love to avoid DOM traversal.

File: packages/drawer/src/react/__specs__/storyshots.spec.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import initStoryshots, { snapshotWithOptions } from '@storybook/addon-storyshots'
+
+function createNodeMock(element) {
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/__specs__/storyshots.spec.js]: Since Jest is running in a dom-less environment, it doesn't have dom elements to pass when you have refs in your components. But you can do this to provide mock elements that get passed instead of null.

File: packages/drawer/src/react/collapsible.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
+import core from '@pluralsight/ps-design-system-core'
+import glamorous from 'glamorous'
+import { bool, node, string } from 'prop-types'
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/collapsible.js]: We've found it a bit easier to read when using the destructured types but we can change it if you'd like it to be consistent with the rest of the components.

File: packages/drawer/src/react/collapsible.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import React from 'react'
+
+const Container = glamorous.div({
+  overflow: 'hidden',
+  transition: `height ${core.motion.speedNormal}`,
```
[the remark this answers is not in the document]
Comment (dborodyansky) [reply packages/drawer/src/react/collapsible.js]: Yes 👍

[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/collapsible.js]: We weren't quite sure on this one. Maybe we should ask Dmitry? We could do like you suggested the other day and do `ease-out` when opening and `ease-in` when closing.

File: packages/drawer/src/react/collapsible.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    const autoHeight = window.getComputedStyle(element).height
+    element.style.height = prevHeight
+    this.forceRepaint(element)
+    element.style.height = autoHeight
+  }
+  setTransitionEnabled(enabled, element) {
+    element.style.transition = enabled ? '' : 'none'
+  }
+  forceRepaint(element) {
+    element.offsetHeight
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/collapsible.js]: It forces a repaint for the element so that the css transition starts with the right properties. I learned this trick a while ago from this stack overflow answer: https://stackoverflow.com/a/3485654

File: packages/drawer/src/react/index.js (hunk 2 above)
Comment (KaidenR) [reply packages/drawer/src/react/index.js]: There are a couple display issues that show up only when adding a `Row` component inside the `Drawer`.
- The borders of the row don't extend far enough, so we have to add our own border. In order to not double-up on borders we have to shift the row up a single pixel.
- The padding on the left of the DrawerBase is really only necessary when using a `Row`. So we're using RowContainer to add that spacing.

Basically, we're planning on saying that if you need to use a `Row` in the `Drawer`, you should wrap it in a `RowContainer`. Otherwise you won't need to use it.

We thought this was the best solution for the consumer but would love to hear your feedback and suggestions..

File: packages/drawer/src/react/index.js (hunk 1 above)
Comment (KaidenR) [reply packages/drawer/src/react/index.js]: It's the padding needed to push the right-most action button in the `Row` to the correct position in the mocks. We're actually tweaking this and it'll probably change. We'd love to be able to get this to a `core.layout.spacing` number.

File: packages/drawer/src/react/index.js (hunk 3 above)
Comment (KaidenR) [reply packages/drawer/src/react/index.js]: Yeah we're pulling in a polyfill (`element-closest`) at the top of this file. This seemed like the simplest way to ignore anchor and button tag clicks but we're totally open to other suggestions.

File: packages/drawer/src/react/collapsible.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  }
+  async waitForHeightTransitionToEnd(element) {
+    return new Promise(resolve => {
+      element.addEventListener('transitionend', function transitionEnd(event) {
```
[the remark this answers is not in the document]
Comment (bressain) [reply packages/drawer/src/react/collapsible.js]: We don't think so? We've had this code in production for about a year now and haven't had a problem. We're not sure what we would reject on.

File: packages/drawer/.storybook/.babelrc [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (jaketrent) [inline packages/drawer/.storybook/.babelrc]: Is this required?  Storybook config is based on CRA already.

File: packages/drawer/src/react/collapsible.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+      this.containerElement.style.height = 0
+    }
+  }
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.isOpen !== this.props.isOpen) {
+      this.toggle(nextProps.isOpen)
+    }
+  }
+
+  toggle(isOpen) {
```
Comment (jaketrent) [inline packages/drawer/src/react/collapsible.js]: The toggle, open, and close stuff is complicated!  I'm glad you brought your experience and time with this problem to help the greater system.  This is exactly the kind of stuff that we would have missed without your help! Thank you.

Comment (jaketrent) [reply packages/drawer/src/react/collapsible.js]: In fact, I would love to just sit with you and have you verbally walk through some of these lines so I can understand it better and what it helps solve.  Would that be ok?

File: packages/drawer/stories/index.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
+import { action } from '@storybook/addon-actions'
```
Comment (jaketrent) [inline packages/drawer/stories/index.js]: I'm getting this error when running storybook:

```
[glamor] an unexpected rule cache miss occurred. This is probably a sign of multiple glamor instances in your app. See https://github.com/threepointone/glamor/issues/79
```

This usually means that I've installed glamor in 2 places.  I doesn't look like it's in the drawer/package.json.  I've uninstalled/reinstalled everything, still see the prob.  Did/are you seeing this?

File: packages/drawer/.storybook/.babelrc (hunk 9 above)
Comment (KaidenR) [reply packages/drawer/.storybook/.babelrc]: We needed for something else that didn't pan out so yeah, don't need it anymore.

File: packages/drawer/src/react/collapsible.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    const autoHeight = window.getComputedStyle(element).height
+    element.style.height = prevHeight
+    this.forceRepaint(element)
+    element.style.height = autoHeight
+  }
+  setTransitionEnabled(enabled, element) {
+    element.style.transition = enabled ? '' : 'none'
+  }
+  forceRepaint(element) {
+    element.offsetHeight // see https://stackoverflow.com/a/3485654
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/collapsible.js]: This code has been used in Channels for over a year now and we haven't ever seen or heard of any issues with it. The Stack Overflow post mentions this is only a fix for an rendering issue on Webkit. The other browsers seem to render things fine without it and are not affected by doing it.

File: packages/drawer/src/react/collapsible.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  setContainerElement = e => {
+    this.containerElement = e
+  }
+
+  render() {
+    return <Container innerRef={this.setContainerElement}>{this.props.children}</Container>
+  }
+}
+
+Collapsible.displayName = 'Collapsible'
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/collapsible.js]: Okay sweet, we'll remove it.

File: packages/drawer/src/react/index.js [hunk 14]
```diff
@@ -1,19 +1,23 @@
...
 import * as glamor from 'glamor'
 import glamorous, { Div } from 'glamorous'
 import { transparentize } from 'polished'
 import Icon, { sizes as iconSizes } from '@pluralsight/ps-design-system-icon/react'
 import { names as themeNames } from '@pluralsight/ps-design-system-theme/react'
-import { bool, func, node, string } from 'prop-types'
+import PropTypes from 'prop-types'
 import React from 'react'
 
+if (typeof window !== 'undefined')
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/index.js]: Yes. We were getting an error in the build because it was trying to access window in a node environment.

File: packages/drawer/src/react/index.js [hunk 15]
```diff
@@ -36,26 +40,27 @@ const DrawerPanel = glamorous.div(
     transition: `box-shadow ${core.motion.speedNormal}`
   },
   ({ isOpen, themeName }) =>(isOpen
-    ? { boxShadow: `inset 0 1px 3px 0 rgba(0,0,0,${themeName === themeNames.dark ? '0.5' : '0.1'})` }
+    ? { boxShadow: `inset 0 1px 3px 0 rgba(0,0,0,${themeName === themeNames.light ? '0.1' : '0.5'})` }
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/index.js]: Exactly. The only difference is the default. If they don't provide anything we should default to dark.

File: packages/drawer/src/react/index.js [hunk 16]
```diff
@@ -36,26 +40,27 @@ const DrawerPanel = glamorous.div(
...
     : {}
   )
 )
 
 const RowContainerWrapper = glamorous.div(
+  {
+    marginRight: -CARET_AREA_WIDTH,
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/index.js]: In my experience units have always defaulted to `px` if not provided. Bressain says he's only run into `em`s being used for font-related properties.

File: packages/drawer/src/react/index.js [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+      ':hover': {
+        background: transparentize(0.8, gray02)
+      }
+    }
+    : {}
+)
+
+const DrawerPanel = glamorous.div(
+  {
+    background: transparentize(0.94, gray02),
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/index.js]: screen shot 2017-11-21 at 2 44 09 pm

File: packages/drawer/src/react/index.js [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+  render() {
+    const { isOpen, context: { themeName } } = this
+    return (
+      <div>
+        <DrawerBase isOpen={isOpen} onClick={this.handleClick}>
+          {this.props.base}
+          <ToggleButtonContainer>
+            <ToggleButton onClick={this.toggle} themeName={themeName}>
+              <RotatableIcon id="caretDown" isRotated={isOpen} />
```
[the remark this answers is not in the document]
Comment (KaidenR) [reply packages/drawer/src/react/index.js]: RotatableIcon is just a glamorous wrapper around the DS Icon component so `id` is just passed through to it. We'll switch it over to use `Icon.ids.caretDown` though.

File: packages/drawer/src/react/index.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (jaketrent) [reply packages/drawer/src/react/index.js]: Yeah, that makes sense.  Sounds good to me.  I understand wanting to be a number so you can do math on it.  Thusfar, all in-js css values are px strings if the math can be accomplished via a `calc(${string})` css string.

File: packages/drawer/src/react/index.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (jaketrent) [reply packages/drawer/src/react/index.js]: Haha, awesome.  Thanks for helping me out.  (I wasn't looking at the light theme).

File: packages/drawer/src/react/index.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (jaketrent) [reply packages/drawer/src/react/index.js]: Ah yes, using the enum is definitely suggested, but through the `ids` export.

File: packages/drawer/src/react/index.js (hunk 2 above)
Comment (jaketrent) [reply packages/drawer/src/react/index.js]: We chatted as a group and decided to make the Drawer support the Row.  We don't know what else it should/will support someday.  We'll optimize the primary case.  RowContainer should be able to go away.  Margin -1 still needs to happen.

File: packages/drawer/stories/index.js (hunk 11 above)
Comment (KaidenR) [reply packages/drawer/stories/index.js]: After lots of comment-half-the-code-out testing, we found that the error was coming from our use of the `glamorous` function to wrap the Icon component. When we took that out, everything worked fine. So we refactored our code to not use it and aren't seeing the error anymore.

Comment (jaketrent) [reply packages/drawer/stories/index.js]: Wow, that is nuts!  And still doesn't make sense to me.  When you get a moment, will you please point me to the line you changed for my future reference?

File: packages/drawer/src/react/index.js (hunk 3 above)
Comment (jaketrent) [reply packages/drawer/src/react/index.js]: Group decided that element-closest is the most straightforward solution.

File: packages/drawer/src/react/collapsible.js (hunk 10 above)
Comment (jaketrent) [reply packages/drawer/src/react/collapsible.js]: done.  thanks for the extra info.
